test/test_charm.py

```diff
@@ -95,6 +95,30 @@ def _on_start(self, event):
         with self.assertRaisesRegex(TypeError, "observer methods must now be explicitly provided"):
             framework.observe(charm.on.start, charm)
 
+    def test_observe_decorated_method(self):
+        events = []
+
+        def dec(fn):
+            @functools.wraps(fn)
+            def wrapper(charm, evt):
+                events.append(evt)
+                fn(charm, evt)
+            return wrapper
+
+        class MyCharm(CharmBase):
+            def __init__(self, *args):
+                super().__init__(*args)
+                framework.observe(self.on.start, self._on_start)
+
+            @dec
+            def _on_start(self, event):
+                pass
+
+        framework = self.create_framework()
+        charm = MyCharm(framework)
+        charm.on.start.emit()
+        self.assertEqual(1, len(events))
```

Comment:
Since it's not obvious what exactly we're trying to test from looking at the code (at least for me), I really think some good in-line code comments here would be invaluable.  Explaining that we're trying to ensure we don't do anything in ops that breaks the ability to wrap hook functions with decorators (at least using functools since we've already broken plain decorators).